### PR TITLE
Port historical schemas to jsonschema

### PR DESCRIPTION
**Pull request by rw-access**
## Issues
No issue, just continuing to clean things up.

## Summary
I moved historical schemas outside of JSL and instead saved them as a jsonschema file. One per type.
Also added a method to dump the current schema in the jsonschema form.

This is only for the Kibana API payload. This way, we'll only ever have one schema in TOML form.
The downgrade workflow still works but reads from those files instead.

Added a new command `python -m detection_rules dev update-schemas` to save the active dataclass to `master/master.<type>.json`.

### CODE REVIEW

File: detection_rules/schemas/__init__.py [hunk 1]
```diff
@@ -2,64 +2,141 @@
...
+        assert version not in migrations
+        migrations[version] = f
+        return f
+
+    return wrapper
+
+
+@migrate("7.8")
+@migrate("7.9")
+@migrate("7.12")
```
Comment (brokensound77) [inline detection_rules/schemas/__init__.py:45]: it feels error-prone to remember to add a decorator for each new version. 
 Can you build it dynamically for any schema version which does not have an explicit override.

File: detection_rules/schemas/definitions.py [hunk 2]
```diff
@@ -10,6 +10,8 @@
 from marshmallow import validate
 from marshmallow_dataclass import NewType
 
+CURRENT_STACK_VERSION = "7.13.0"
```
Comment (brokensound77) [inline detection_rules/schemas/definitions.py:13]:
```python
Package.load_configs()['name']
```

File: detection_rules/schemas/__init__.py [hunk 3]
```diff
@@ -2,64 +2,141 @@
...
+def strip_additional_properties(version: Version, api_contents: dict) -> dict:
+    """Remove all fields that the target schema doesn't recognize."""
+    rule_type = api_contents["type"]
+    stack_dir = Path(SCHEMA_DIR) / str(version)
+    schema_file = stack_dir / f"{version}.{rule_type}.json"
+
+    if not schema_file.exists():
+        raise ValueError(f"Unsupported rule type {rule_type}")
+
+    target_schema = json.loads(schema_file.read_text(encoding="utf8"))
```
Comment (brokensound77) [inline detection_rules/schemas/__init__.py:55]: should these loads be cached, at least for the duration of the `downgrade` function

**Review by brokensound77 — COMMENTED** [review]
This is pretty slick. Left some comments. Could benefit from adding an ability in the cli to validate a rule against a specified version

**Comment by brokensound77** [comment]
> marshmallow.exceptions.ValidationError: {'rule': [ValidationError({'type': ['Must be equal to query.'], 'threat_indicator_path': ['Unknown field.']

File: detection_rules/devtools.py [hunk 4]
```diff
@@ -702,3 +702,12 @@ def validate_ml_detections_asset(directory):
     click.echo(f'description to paste with release:\n\n{description_str}\n')
 
     return zip_name, description_str
+
+
+@dev_group.command("update-schemas")
+def update_schemas():
+    from . import rule
+    classes = [rule.BaseQueryRuleData, rule.MachineLearningRuleData, rule.ThresholdQueryRuleData, rule.EQLRuleData]
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/devtools.py]: d2138d4d6761d9d78b7b4af8b00c4f7f8bb5f185

File: detection_rules/cli_utils.py [hunk 5]
```diff
@@ -111,9 +112,10 @@ def rule_prompt(path=None, rule_type=None, required_only=True, save=True, verbos
         kwargs.update(kwargs.pop('rule'))
 
     rule_type = rule_type or kwargs.get('type') or \
-        click.prompt('Rule type', type=click.Choice(CurrentSchema.RULE_TYPES))
+        click.prompt('Rule type', type=click.Choice(typing.get_args(definitions.RuleType)))
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/cli_utils.py:115]: d2138d4d6761d9d78b7b4af8b00c4f7f8bb5f185

File: detection_rules/rule.py [hunk 6]
```diff
@@ -291,7 +307,7 @@ def validate_query(self, meta: RuleMeta) -> None:
 
 
 # All of the possible rule types
-AnyRuleData = Union[QueryRuleData, EQLRuleData, MachineLearningRuleData, ThresholdQueryRuleData, ThreatMatchRuleData]
+AnyRuleData = Union[QueryRuleData, EQLRuleData, MachineLearningRuleData, ThresholdQueryRuleData]
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/rule.py]: d2138d4d6761d9d78b7b4af8b00c4f7f8bb5f185

File: detection_rules/rule.py [hunk 7]
```diff
@@ -300,6 +316,25 @@ class TOMLRuleContents(MarshmallowDataclassMixin):
     metadata: RuleMeta
     data: AnyRuleData = field(metadata=dict(data_key="rule"))
 
+    @classmethod
+    def all_rule_types(cls):
+        types = set()
+        for subclass in typing.get_args(AnyRuleData):
+            field = next(field for field in dataclasses.fields(subclass) if field.name == "type")
+            types.update(typing.get_args(field.type))
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/rule.py:324]: it's a `set()`

File: detection_rules/schemas/definitions.py (hunk 2 above)
Comment (rw-access) [reply detection_rules/schemas/definitions.py:13]: 33114c98f367468ed5b718b911c7c148e8b2ac6a

Comment (rw-access) [reply detection_rules/schemas/definitions.py]: 33114c98f367468ed5b718b911c7c148e8b2ac6a

File: detection_rules/schemas/__init__.py [hunk 8]
```diff
@@ -2,64 +2,141 @@
...
 
 
-def downgrade(api_contents: dict, target_version: str):
-    """Downgrade a rule to a target stack version."""
-    # truncate to (major, minor)
-    target_version_str = target_version
-    target_version = Version(target_version)[:2]
-    versions = set(Version(schema_cls.STACK_VERSION) for schema_cls in all_schemas)
-    role = api_contents.get("type")
+def all_versions():
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/schemas/__init__.py]: ba151bb9a4bda9a2b85787bbbbaeeb0902176bee

File: detection_rules/schemas/__init__.py [hunk 9]
```diff
@@ -2,64 +2,141 @@
...
+    """Get all known stack versions."""
+    return [str(v) for v in sorted(migrations)]
+
+
+def migrate(version):
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/schemas/__init__.py]: ba151bb9a4bda9a2b85787bbbbaeeb0902176bee

File: detection_rules/rule.py [hunk 10]
```diff
@@ -300,6 +316,25 @@ class TOMLRuleContents(MarshmallowDataclassMixin):
...
+
+        return types
+
+    @classmethod
+    def get_data_subclass(cls, rule_type):
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/rule.py]: ba151bb9a4bda9a2b85787bbbbaeeb0902176bee

File: detection_rules/schemas/__init__.py [hunk 11]
```diff
@@ -2,64 +2,141 @@
...
-all_schemas = [
-    ApiSchema78,
-    ApiSchema79,
-    ApiSchema710,
-    ApiSchema711,
-    ApiSchema712,
-]
-CurrentSchema = all_schemas[-1]
-available_versions = [cls.STACK_VERSION for cls in all_schemas]
+SCHEMA_DIR = Path(__file__).absolute().parent.parent.parent / "etc" / "api_schemas"
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/schemas/__init__.py]: ba151bb9a4bda9a2b85787bbbbaeeb0902176bee

**Review by threat-punter — APPROVED** [review]

**Comment by brokensound77** [comment]
TODO: add migration for ML in 7.13 here or new PR after this merges (ref #1167)

File: detection_rules/schemas/__init__.py (hunk 1 above)
Comment (brokensound77) [reply detection_rules/schemas/__init__.py:58]: any thoughts on this?

**Comment by brokensound77** [comment]
Every new release, we bump the version (`name`) in `packages.yml`. Should we add a unit test to validate a flat json schema is preserved once a bump is made - or maybe just a command that saves then bumps

File: detection_rules/schemas/__init__.py (hunk 3 above)
Comment (rw-access) [reply detection_rules/schemas/__init__.py]: addressed a while ago (Outdated PR comment) but here's the commit 1dd56d7fe6721fed9c4336e2573e05b81720595f

File: detection_rules/packaging.py [hunk 12]
```diff
@@ -549,3 +549,8 @@ def create_bulk_index_body(self) -> Tuple[Ndjson, Ndjson]:
             importable_rules_docs.append(rule_doc)
 
         return bulk_upload_docs, importable_rules_docs
+
+
+@cached
+def current_stack_version() -> str:
```
[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/packaging.py:555]: can you explain and give a recommendation what you would change?

[the remark this answers is not in the document]
Comment (brokensound77) [reply detection_rules/packaging.py:555]: choose one function and stick with it. Either this or `load_current_package_version`. Probably keep this one and replace references to the other with this one

[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/packaging.py:555]: what is `load_current_package_version` that you speak of?
L70 is `load_versions`

[the remark this answers is not in the document]
Comment (rw-access) [reply detection_rules/packaging.py:555]: oh. I see what happened

**Pull request merged**